_posts/2021-09-XX-ansible-dokku.md

```diff
@@ -0,0 +1,401 @@
+
+# app-level resource limits
+my_dokku_cpu_limit: 2  # max. number of CPUs per app
+my_dokku_memory_limit: 4G # max. memory per app in GBytes
+
+dokku_users:
+  - username: ltalirz
+    email: lt@gmail.com
+    ssh_key: "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQChkphp/4wgTnVrYNuyE6eVNcMFzOZwglj4cP8IAtZwEde8y3KwKEHX6jG3j2tFiV9EGS+7aLVqXsHf/OFTHjad54OEafG5/BR/c2dVhk7dOGhPTCvZza8fX/9IyZiYRRBGx2lH4Y9ypSJyQ/6QzpgCe0Qd9zqnVWrqyY4ny+vsuGiSp+QWPNs0YD+tmvmPh1RvMxWdWMKGXt1QLscN3azRUl140wWLK3uAc85tqgyGCAFxOK8rOpok8GoypD2CRBNfyyislUWE+OGXxdhwOwaYEbj9vScTiMDPuEDgXpoKEaEPnMh5axmy38IXOoNhnndwwLzqiIs6ssiYUOjvmqgD"
+    apps:
+      - name: seekpath
+      - name: oximachine
+      - name: mofchecker
+```
+
+The list contains, for each user:
+ * A `username` used as a dokku-internal identifier (just pick one)
+ * A contact `email` (optional)
+ * The user's public SSH key
+ * A list of `apps` the user is allowed to manage
+
+In order to provision our dokku instance accordingly, we need to loop twice:
+first over the list of users and then over the list of apps.
+ 
+Let's add a task to our `playbook.yml`:
+```yaml
+- name: configure users
+  - name: Create apps for dokku users
+    include_tasks: user-single.yml
+    vars:
+      dokku_user: "{{ item }}"
+    with_items: "{{ dokku_users }}"
+    tags: my_dokku_users
+```
+
+Here we are looping over the users, running `app-single.yml` for each user.
+
+`user-single.yml`: ansible task list that provisions apps for a single user
+```yaml
+---
+- name: List apps for user "{{ dokku_user.username }}"
+  command: dokku acl:allowed {{ dokku_user.username }}
+  register: allowed_apps
+
+- name: Process apps for user "{{ dokku_user.username }}"
+  include_tasks: app-single.yml
+  loop_control:
+    loop_var: dokku_app
+  with_items: "{{ dokku_user.apps }}"
+```
+
+And finally, for each app listed we will (1) create the app, (2) give the user access to manage the app, and (3) apply the resource limits to it.
+For the moment, resource limits are global but the approach above illustrates how they could be adapted per app.
+
+`app-single.yml`: ansible task list that provisions a single app for a user
+```yaml
+---
+- name: Create app {{ dokku_app.name }}
+  dokku_app:
+    app: "{{ dokku_app.name }}"
+  when: dokku_app.name not in allowed_apps.stdout_lines
+
+- name: Let user {{ dokku_user.username }} manage app {{ dokku_app.name}}
+  command: dokku acl:add {{ dokku_app.name }} {{ dokku_user.username }}
+  when: dokku_app.name not in allowed_apps.stdout_lines
+
+- name: Set default resource limits for app {{ dokku_app.name}}
+  command: >
+    dokku resource:limit \
+      --memory {{ my_dokku_memory_limit }} \
+      --cpu {{ my_dokku_cpu_limit }} \
+      {{ dokku_app.name }}
+```
+
+The following will create the new apps and give the users access to them:
+
+```
+ansible-playbook playbook.yml --tags my_dokku_users
+```
+
+
+Finally, let's add some global configuration concerning which commands regular users are allowed to run
```

Comment:
I'd like to avoid any references to plugins that aren't directly maintainable by the ansible modules if possible. That goes for the above as well where you modify acl + change resources.

Reply:
To me, the ability to give non-admin users access to individual apps is a core feature of any PaaS.
Is there a way to do it in dokku without the `acl` plugin?

If not, I think it may still be worth including it here.
We can even mention this example in the conclusion section and point it out as something left to "modularize".

Reply:
I think maybe a separate tutorial would be better?

Aside, I have an idea around how to make auth "native", but its definitely just an idea: https://gist.github.com/josegonzalez/0f8b517a560e705b122d5a1531c1aa22

Reply:
As you prefer - to me, provisioning dokku just with app or two would be using it as a platform but missing the "aaS" part

Reply:
of course, the alternative is to just add a `dokku_acl` module - should be very straightforward.

if I find the time I can give it a look

Reply:
I've now created a `dokku_acl_app` module and am using it to set acls on the apps